File: haskell/glue/src/Glue/Lib/Builtin.hs
module Glue.Lib.Builtin where

import Glue.Eval (Eval)
import Glue.Lib.Builtin.At (at)
import Glue.Lib.Builtin.Def (def)
import Glue.Lib.Builtin.Error (errorFunc)
import Glue.Lib.Builtin.Lambda (lambda)
import Glue.Lib.Builtin.Let (let')
import Glue.Lib.Builtin.List (list)
import Glue.Lib.Builtin.Quote (quote)
import Glue.Lib.Builtin.Try (tryFunc)
import Glue.Lib.Builtin.Void (void)
import Glue.Module (ModuleInfo, nativeModule)

builtin :: ModuleInfo Eval
builtin =
    nativeModule
        "ffi.builtin"
        [ ("def", def)
        , ("lambda", lambda)
        , ("\\", lambda)
        , ("let", let')
        , ("error", errorFunc)
        , ("try", tryFunc)
        , ("quote", quote)
        , ("list", list)
        , ("#", list)
        , ("void", void)
        , ("_", void)
        , ("at", at)
        , ("@", at)
        ]

File: dart/glue/lib/src/lib/builtin.dart
import 'package:glue/src/lib/builtin/def.dart';
import 'package:glue/src/lib/builtin/error.dart';
import 'package:glue/src/lib/builtin/import.dart';
import 'package:glue/src/lib/builtin/lambda.dart';
import 'package:glue/src/lib/builtin/let.dart';
import 'package:glue/src/lib/builtin/quote.dart';
import 'package:glue/src/lib/builtin/try.dart';
import 'package:glue/src/module.dart';

/// Builtin module - special forms and core language constructs
/// Mirrors Haskell Glue.Lib.Builtin exactly

/// The builtin module containing all special forms
/// Mirrors Haskell Glue.Lib.Builtin.builtin exactly
final ModuleInfo builtinModule = nativeModule('ffi.builtin', [
  ('def', def),
  ('lambda', lambda),
  ('\\', lambda),
  ('let', let),
  ('import', importForm),
  ('error', error),
  ('try', tryFunc),
  ('quote', quote),
]);
